### PR TITLE
Properly tear down ORC ExecutionSession

### PR DESCRIPTION
**Pull request by alexbaden**
Makes a few changes to resolve a leak discovered in the way our ORC ExecutionSession / ExecutionSessionWrapper teardown process worked. Mainly, we now call `endSession` on the `orc::ExecutionSession`. To do this correctly, we need to delete the move constructors for `ExecutionSessionWrapper` -- otherwise the move will call the destructor which will purge resources too soon. This does mean we need to unique_ptr the wrapper in the Backend code. I tried a bunch of other permutations where I avoided doing this, but it ended up introducing too much additional code and bugs to avoid one heap allocation. 

Now that we are calling `endSession`, some previous hacks are no longer necessary. For example, the `SectionMemoryManager` is cleaned up with `RTDyldObjectLinkingLayer`, which is cleaned up during `endSession` -- so we remove the custom memory manager class. I also removed some constructors which were not needed.

### CODE REVIEW

File: QueryEngine/CompilationContext.h [hunk 1]
```diff
@@ -27,23 +27,25 @@ class CompilationContext {
...
 
   void* getPointerToFunction(llvm::Function* function) {
-    return execution_engine_.getPointerToFunction(function);
+    return execution_engine_->getPointerToFunction(function);
   }
 
-  void* func() const { return func_; }
+  void* func() const {
+    return func_;
+  }
```
Comment (kurapov-peter) [inline QueryEngine/CompilationContext.h:48]: Is this llvm14 clang format?

**Comment by alexbaden** [comment]
Yes, we might be able to resolve it by changing the flags. I went ahead and formatted in a clang-12 env for now, and also picked up the cuda build fix so this one should be green.

**Review by ienkovich — APPROVED** [review]
Nice fix! Does it fix all current ASAN errors?

**Comment by alexbaden** [comment]
It did for `ArrowBasedExecuteTest` in my env.

**Pull request merged**